analytics/analytics/core/management/commands/migrate_exit_code_and_failure_reasons.py

```diff
@@ -0,0 +1,123 @@
+from itertools import batched
+
+from django.db import connections
+import djclick as click
+
+from analytics.core.models import JobDataDimension
+
+
+def _migrate_exit_codes(dry_run: bool) -> None:
+    # Only failed jobs have an exit code in the database
+    failed_jobs = JobDataDimension.objects.filter(status="failed")
+
+    # Batch the job ids to avoid hitting the max query length/memory limit
+    BATCH_SIZE = 5_000
+    for job_ids in batched(
+        failed_jobs.values_list("job_id", flat=True).iterator(chunk_size=BATCH_SIZE), BATCH_SIZE
+    ):
+        with connections["gitlab"].cursor() as cursor:
+            QUERY = "SELECT build_id, exit_code FROM public.ci_builds_metadata WHERE build_id IN %s"
+            cursor.execute(QUERY, [tuple(job_ids)])
+            rows = cursor.fetchall()
+
+        job_id_to_exit_code_mapping = {row[0]: row[1] for row in rows if row[1] is not None}
+
+        if dry_run:
+            click.echo(f"Would update {len(job_id_to_exit_code_mapping.keys())} jobs")
+            continue
+
+        click.echo(f"Updating {len(job_id_to_exit_code_mapping.keys())} jobs")
+        jobs_to_update = list(
+            JobDataDimension.objects.filter(job_id__in=job_id_to_exit_code_mapping.keys())
+        )
+        for job in jobs_to_update:
+            job.job_exit_code = job_id_to_exit_code_mapping[job.job_id]
+
+        JobDataDimension.objects.bulk_update(jobs_to_update, ["job_exit_code"])
+
+
+# From https://gitlab.com/gitlab-org/gitlab/-/blob/master/app/models/concerns/enums/ci/commit_status.rb#L7
+FAILURE_REASONS = {
```

Comment:
I wonder if you could use this in the opposite direction to eliminate the need for `get_job_exit_code`? If the mapping is consistent, it could remove the need to query the gitlab DB.

Reply:
This enum mapping is unrelated to the exit code, it's just used to map the `ci_builds.failure_reason` column to a string.